### DISCUSSION TITLE
After install iOS/iPadOS app - integrity cannot be verified

### DISCUSSION THREAD

**Discussion by xpunkt-torie** [airsdk/Adobe-Runtime-Support#2487] [General]
We have been working on a mobile app for some time now. The decision to use Starling and Air was made because of the cross platform possibilities and because of the existing knowledge in AS3.

Initially Flash Builder now VS Code was used as IDE on MacOS. The app is still in development (for now we are focusing on iOS and iPadOS). Testing and debugging with VS Code works very well. (Except for a problem with adl on Retina MacBooks.) Testing by installing on devices has also worked without problems so far. But now we had to recreate the certificates and profiles at Apple Developer and include them in asconfig.json. We think we did this as usual. But unfortunately it does not work anymore.

We can build the ipa without problems and load it via Configurator to the iOS/iPadOS device (also the newProvisioning Profile). But the icon on the e.g. iPad has an additional icon, a small cloud/download icon. When trying to start the app on the iPad we get this error (German message - translated):

```
"appname" cannot be installed.
This app cannot be installed because its integrity could not be verified.
```

We don't know what we could have done wrong. So far, everything has worked great. We are happy to provide the individual steps of the certificate and profile creation if that should be of interest.

Here is our configuration:
MacBookPro M1 with macOS 13.2
Visual Studio Code 1.75.0
ActionScript & MXML v1.15.0
AIR SDK 50.1.1.2 / 33.1.1.889
iOS / iPadOS 16.3.1 (with Developer Mode enabled) / 15.x.x
Is there a step by step guide to create the certificates and profiles for our case? We want to install/use the app for testing on different devices, all of which will be listed on Apple and included in the profile.

Thanks in advance.

**Comment by ajwfrost** [comment]
I suspect `its integrity could not be verified` means that something is up with the code signature... although if you were using a macbook, I would have hoped it's using the Apple code-signing and hence should be watertight....

But could you please take the .ipa file and unzip it, you'll get a folder called "Payload" and then the application itself `appname.app`. If you go to this in a terminal window, try running:
```
codesign -v -vvv appname.app
```
and
```
codesign -d -vvv appname.app
```
and we can see if that highlights anything..

thanks

**Comment by xpunkt-torie** [comment]
Yes, we develop on a MacBook Pro M1.
The result of the two codesign calls is here:

```
~ % codesign -v -vvv APPNAME.app
/Users/torie/xpunkt-dev/APPNAME/bin/Payload/APPNAME.app: a sealed resource is missing or invalid
file missing: /Users/torie/xpunkt-dev/APPNAME/bin/Payload/APPNAME.app/.DS_Store

~ % codesign -d -vvv APPNAME.app
Executable=/Users/torie/xpunkt-dev/APPNAME/bin/Payload/APPNAME.app/APPNAME
Identifier=com.APPNAME.APPNAME
Format=app bundle with Mach-O thin (arm64)
CodeDirectory v=20400 size=103099 flags=0x0(none) hashes=3211+7 location=embedded
Hash type=sha256 size=32
CandidateCDHash sha1=86c7b3384bb5b95eca13d97a056c5709d24eff6a
CandidateCDHashFull sha1=86c7b3384bb5b95eca13d97a056c5709d24eff6a
CandidateCDHash sha256=f1bcc7a98b2b69b6c2df30b382edf9a2795c2c65
CandidateCDHashFull sha256=f1bcc7a98b2b69b6c2df30b382edf9a2795c2c6520c78f700e39239ff9c3bd2e
Hash choices=sha1,sha256
CMSDigest=f172c39c11c11b4e7c63a7118473455f8f1e12d2214940a4af12ec0155cee70d
CMSDigestType=2
Launch Constraints:
	None
CDHash=f1bcc7a98b2b69b6c2df30b382edf9a2795c2c65
Signature size=4842
Authority=iPhone Developer: MY NAME (73H3UT4JZ3)
Authority=Apple Worldwide Developer Relations Certification Authority
Authority=Apple Root CA
Signed Time=26. Feb 2023 at 15:54:28
Info.plist entries=52
TeamIdentifier=G9C93K59YC
Sealed Resources version=2 rules=10 files=338
Internal requirements count=1 size=188
```

I hope this gives further helpful insights.

Thanks in advance.

**Reply by ajwfrost** [reply]
Hi

Yes that's shown what the issue is:
```
a sealed resource is missing or invalid
file missing: APPNAME.app/.DS_Store
```

I suspect then, when we did the code-signing, there was a .DS_Store file present in the folder structure, but when the packaging of the IPA file was done, this file was ignored. Can you make sure there are no such files present in your build folders, it could be brought in if you'd packaged up a folder with a wildcard (e.g. `adt ... -C bin .` would pick up all files in the `bin` folder, including hidden ones like this).

Unless you've changed your Finder settings, the simplest way to find these files might be via a terminal window, using `find /your/application/folder/ -name ".DS_Store"`

thanks

**Reply by xpunkt-torie** [reply]
Hi 

Yes, I found ".DS_Store" files in almost all folders of the project. Unfortunately, they really don't show up in the Finder anymore, even if you get hidden files displayed.

I have deleted all ".DS_Store" files in the project:
`
find /your/application/folder/ -name '*.DS_Store' -type f -delete
`

After deleting all ".DS_Store" files in the project, a "DS_Store" file is created again when building in the "bin-debug" folder. 

Test with the new .ipa (.app):
```
~ % codesign -v -vvv /Users/torie/xpunkt-dev/APPNAME/bin/Payload/APPNAME.app 
/Users/torie/xpunkt-dev/APPNAME/bin/Payload/APPNAME.app: valid on disk
/Users/torie/xpunkt-dev/APPNAME/bin/Payload/APPNAME.app: satisfies its Designated Requirement

~ % codesign -d -vvv /Users/torie/xpunkt-dev/APPNAME/bin/Payload/APPNAME.app 
Executable=/Users/torie/xpunkt-dev/APPNAME/bin/Payload/APPNAME.app/APPNAME
Identifier=com.APPNAME.APPNAME
Format=app bundle with Mach-O thin (arm64)
CodeDirectory v=20400 size=103099 flags=0x0(none) hashes=3211+7 location=embedded
Hash type=sha256 size=32
CandidateCDHash sha1=a40286d835aed8e95c91761aa4c921b3f61d2bc4
CandidateCDHashFull sha1=a40286d835aed8e95c91761aa4c921b3f61d2bc4
CandidateCDHash sha256=7fa025a4d106887cddd2ec32e00b5b257d76adf2
CandidateCDHashFull sha256=7fa025a4d106887cddd2ec32e00b5b257d76adf26c193ce5009e51b9b238e24e
Hash choices=sha1,sha256
CMSDigest=29a42ae427a15a8e996f7bbfda609de8dc39d398399f0fc05d7ba535c5f332e5
CMSDigestType=2
Launch Constraints:
	None
CDHash=7fa025a4d106887cddd2ec32e00b5b257d76adf2
Signature size=4842
Authority=iPhone Developer: MY NAME (73H3UT4JZ3)
Authority=Apple Worldwide Developer Relations Certification Authority
Authority=Apple Root CA
Signed Time=27. Feb 2023 at 10:00:55
Info.plist entries=52
TeamIdentifier=G9C93K59YC
Sealed Resources version=2 rules=10 files=337
Internal requirements count=1 size=188
```
The .ipa can now be installed and started on the iPad again 😀.

I'm surprised this problem hasn't come up here, not on the Starling forum or other forums yet. That with the "DS_Store" files affects aale developers on macOS after all.

Since codesign comes from Apple itself, I am surprised that Apple Developer Support, which I had already described my problem in detail, has not suggested these tests. Probably Apple would prefer that you develop with XCode. Who knows.

@ajwfrost Thank you very much for the quick help.

Thanks to the great community.

**Reply by ajwfrost** [reply]
As an interesting aside, I'm unable to create an IPA file that has this problem.. I've been trying to add .DS_Store files into the path and include them in the package but (a) they're not in the package, and (b) they're not in the code signature either, regardless of whether I get it to use macOS Codesign or the internal ADT signing logic...

So I'm not quite sure how you ended up with this situation! Seems a strange one..

**Reply by xpunkt-torie** [reply]
Yes, it's all very strange.
I have now searched again for ".DS_Store" files in the project folder and found them.
```
~ % find /Users/torie/xpunkt-dev/APPNAME -name ".DS_Store"
/Users/torie/xpunkt-dev/APPNAME/.DS_Store
/Users/torie/xpunkt-dev/APPNAME/bin/.DS_Store
/Users/torie/xpunkt-dev/APPNAME/bin-debug/.DS_Store

```
However, these do not seem to have any influence on the problem, since it now works anyway. Both the check with "codesign" and the installation of the app on the device.

Probably this must have been a "DS_Store" file in another folder in the project.
